File: implementations/rust/src/b64u.rs
//! Canonical unpadded base64url (RFC 4648 §5) — the JSON-profile transport for a bytes target's
//! payload (SPEC-1 §4.2, ERRATA D12). Mirrors ../ts/src/b64u.ts. Identity is computed from the raw
//! bytes, never from this encoding; encoding is canonical by construction and decoding *validates*
//! canonicality (reject, never repair — SPEC-4 §2): no padding, alphabet only, length never ≡ 1
//! (mod 4), and a final character's unused low bits MUST be zero.

const ALPHABET: &[u8; 64] = b"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

/// Encode raw bytes as canonical unpadded base64url.
pub fn encode(bytes: &[u8]) -> String {
    let mut out = String::with_capacity(bytes.len().div_ceil(3) * 4);
    for chunk in bytes.chunks(3) {
        let b0 = chunk[0] as u32;
        let b1 = *chunk.get(1).unwrap_or(&0) as u32;
        let b2 = *chunk.get(2).unwrap_or(&0) as u32;
        let n = (b0 << 16) | (b1 << 8) | b2;
        out.push(ALPHABET[(n >> 18) as usize & 63] as char);
        out.push(ALPHABET[(n >> 12) as usize & 63] as char);
        if chunk.len() > 1 {
            out.push(ALPHABET[(n >> 6) as usize & 63] as char);
        }
        if chunk.len() > 2 {
            out.push(ALPHABET[n as usize & 63] as char);
        }
    }
    out
}

fn sextet(c: u8) -> Result<u32, String> {
    let v = match c {
        b'A'..=b'Z' => c - b'A',
        b'a'..=b'z' => c - b'a' + 26,
        b'0'..=b'9' => c - b'0' + 52,
        b'-' => 62,
        b'_' => 63,
        // '=' padding lands here too — deliberately rejected (canonical form is unpadded).
        _ => return Err(format!("base64url: invalid character {:?}", c as char)),
    };
    Ok(v as u32)
}

/// Decode canonical unpadded base64url, rejecting any non-canonical input.
pub fn decode(s: &str) -> Result<Vec<u8>, String> {
    let bytes = s.as_bytes();
    if bytes.len() % 4 == 1 {
        return Err("base64url: invalid length (≡ 1 mod 4)".into());
    }
    let mut out = Vec::with_capacity(bytes.len() / 4 * 3 + 2);
    let mut i = 0;
    while i < bytes.len() {
        let group = &bytes[i..(i + 4).min(bytes.len())];
        let mut acc = 0u32;
        for &c in group {
            acc = (acc << 6) | sextet(c)?;
        }
        match group.len() {
            4 => {
                out.push((acc >> 16) as u8);
                out.push((acc >> 8) as u8);
                out.push(acc as u8);
            }
            3 => {
                // 18 bits carried; the last 2 are spill and MUST be zero.
                if acc & 0x3 != 0 {
                    return Err("base64url: non-canonical trailing bits".into());
                }
                let acc = acc >> 2;
                out.push((acc >> 8) as u8);
                out.push(acc as u8);
            }
            2 => {
                // 12 bits carried; the last 4 are spill and MUST be zero.
                if acc & 0xf != 0 {
                    return Err("base64url: non-canonical trailing bits".into());
                }
                out.push((acc >> 4) as u8);
            }
            _ => unreachable!("chunk of 0 or 1 impossible after the length check"),
        }
        i += 4;
    }
    Ok(out)
}

File: implementations/ts/src/b64u.ts
// Canonical unpadded base64url (RFC 4648 §5) — the JSON-profile transport for a bytes target's
// payload (SPEC-1 §4.2, ERRATA D12). Mirrors ../rust/src/b64u.rs. Identity is computed from the raw
// bytes, never this encoding; encoding is canonical by construction and decoding *validates*
// canonicality (reject, never repair — SPEC-4 §2): no padding, alphabet only, length never ≡ 1
// (mod 4), and a final character's unused low bits MUST be zero.

const ALPHABET = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";

export function b64uEncode(bytes: Uint8Array): string {
  let out = "";
  for (let i = 0; i < bytes.length; i += 3) {
    const b0 = bytes[i]!;
    const b1 = i + 1 < bytes.length ? bytes[i + 1]! : 0;
    const b2 = i + 2 < bytes.length ? bytes[i + 2]! : 0;
    const n = (b0 << 16) | (b1 << 8) | b2;
    out += ALPHABET[(n >> 18) & 63]! + ALPHABET[(n >> 12) & 63]!;
    if (i + 1 < bytes.length) out += ALPHABET[(n >> 6) & 63]!;
    if (i + 2 < bytes.length) out += ALPHABET[n & 63]!;
  }
  return out;
}

function sextet(c: string): number {
  const code = c.charCodeAt(0);
  if (code >= 65 && code <= 90) return code - 65; // A-Z
  if (code >= 97 && code <= 122) return code - 97 + 26; // a-z
  if (code >= 48 && code <= 57) return code - 48 + 52; // 0-9
  if (c === "-") return 62;
  if (c === "_") return 63;
  // '=' padding lands here too — deliberately rejected (canonical form is unpadded).
  throw new Error(`base64url: invalid character ${JSON.stringify(c)}`);
}

export function b64uDecode(s: string): Uint8Array {
  if (s.length % 4 === 1) throw new Error("base64url: invalid length (≡ 1 mod 4)");
  const out: number[] = [];
  for (let i = 0; i < s.length; i += 4) {
    const end = Math.min(i + 4, s.length);
    const len = end - i;
    let acc = 0;
    for (let j = i; j < end; j++) acc = (acc << 6) | sextet(s[j]!);
    if (len === 4) {
      out.push((acc >> 16) & 0xff, (acc >> 8) & 0xff, acc & 0xff);
    } else if (len === 3) {
      // 18 bits carried; the last 2 are spill and MUST be zero.
      if ((acc & 0x3) !== 0) throw new Error("base64url: non-canonical trailing bits");
      const a = acc >> 2;
      out.push((a >> 8) & 0xff, a & 0xff);
    } else {
      // len === 2: 12 bits carried; the last 4 are spill and MUST be zero.
      if ((acc & 0xf) !== 0) throw new Error("base64url: non-canonical trailing bits");
      out.push((acc >> 4) & 0xff);
    }
  }
  return Uint8Array.from(out);
}
